lib/lightdata.h

```diff
@@ -12,17 +12,58 @@ class lightdata {
   unsigned char fine = 0;
   unsigned char tdc = 0;
 
-  int chip() const { return index / 32; };
-  int eoch() const { return index % 64; };
-  int cindex() const { return tdc + 4 * index; };
-  
   lightdata(unsigned char _device, unsigned char _index, unsigned char _coarse, unsigned char _fine, unsigned char _tdc) :
     device(_device),
     index(_index),
     coarse(_coarse),
     fine(_fine),
     tdc(_tdc) { };
-    
+
+  int chip() const { return index / 32; };
+  int eoch() const { return index % 64; };
+  int cindex() const { return tdc + 4 * index; };
+  float time() const;
+
+  /** calibration **/
+
+  static float fine_if[16][768];
+  static float fine_cut[16][768];
+  static float fine_off[16][768];
+  static bool load_fine_calibration(std::string filename);
+
 };
 
+float lightdata::fine_if[16][768] = {0.};
+float lightdata::fine_cut[16][768] = {0.};
+float lightdata::fine_off[16][768] = {0.};
+
+float
+lightdata::time() const
+{
+  auto ci = cindex();
+  float corr = (float)fine * fine_if[device][ci] + fine_off[device][ci];
+  if (fine > fine_cut[device][ci]) corr -= 1.;
+  auto time = (float)coarse - corr;
+  return time;
+}
+
+bool
+lightdata::load_fine_calibration(std::string filename)
+{
+  std::cout << " --- loading fine calibration: " << filename << std::endl;
+  auto fin = TFile::Open(filename.c_str());
+  for (int i = 0; i < 16; ++i) {
+    auto device = 192 + i;
+    auto hFine_if = (TH1 *)fin->Get(Form("hIF_%d", device));
+    auto hFine_cut = (TH1 *)fin->Get(Form("hCUT_%d", device));
+    auto hFine_off = (TH1 *)fin->Get(Form("hOFF_%d", device));
+    for (int j = 0; j < 768; ++j) {
+      fine_if[i][j] = hFine_if ? hFine_if->GetBinContent(j + 1) : 0.;
+      fine_cut[i][j] = hFine_cut ? hFine_cut->GetBinContent(j + 1) : 0.;
+      fine_off[i][j] = hFine_off ? hFine_off->GetBinContent(j + 1) : -fine_if[i][j] * fine_off[i][j];
```

Comment:
This is not offset, I would set it to 0. Moreover: if for some reason you don't have IF and/or CUT this is undefined behaviour

Reply:
To specify: I am referring to -fine_if[i][j] \* **fine_off**[i][j]. I am now realising you wrote fine_off instead of (I believe) fine_cut

Reply:
I dot not understand the comment on OFF.
there is no undefined behaviour here, if the calibration is not existing it will behave like if the calibration is not existing. that is it will not perform any fine correction.

Reply:
fine_off[I][j] is (possibly) undefined, at this point. You haven't reached it yet. if you define, for simplicity, fine_off[I][j] = a and fine_if[i][j] = b you are basically doing the equivalent of:
auto a  = -b*a
Which is textbook undefined behaviour. Moreover I am not clear why you are doing -IF*OFF in the first place.

Reply:
you are right, if must be `fine_cut`
I will correct

Reply:
Then I am unclear why you are doing -IF\*CUT

Reply:
if you have a measurement of OFF, it is the intercept of the linear fit to the "dente di sega" and you use the measurement. 

if you do not have a measurement, it is still the intercept at `fine = 0`.
if you only have IF and CUT you can define OFF just using IF and CUT.